File: packages/routing/src/headers.ts
import { diagnostic } from "./diagnostics.js";
import type { CompiledHeaderOperation, CompiledHeaderRule, RoutingDiagnostic } from "./types.js";
import {
  compilePattern,
  escapeRegex,
  normalizeHostname,
  parseAbsoluteUrl,
  WHITESPACE_PATTERN,
} from "./utils.js";

const HEADER_LINE_LIMIT = 2_000;

// Platform-managed response headers user rules can never set or remove. Mirrors
// PLATFORM_MANAGED_RESPONSE_HEADERS in @spacefast/common static-runtime-policy (parity
// asserted by apps/control-plane php-policy-parity.test.ts); the PHP runtime enforces
// the same map at apply time. The sendfile/accel family is security-critical: under
// nginx+php-fpm those headers turn responses into server-side file reads.
const UNSUPPORTED_RESPONSE_HEADERS = new Set([
  "accept-ranges",
  "age",
  "allow",
  "alt-svc",
  "connection",
  "cookie",
  "content-encoding",
  "content-length",
  "content-range",
  "date",
  "host",
  "location",
  "server",
  "set-cookie",
  "strict-transport-security",
  "trailer",
  "transfer-encoding",
  "upgrade",
  "vary",
  "x-accel-buffering",
  "x-accel-redirect",
  "x-lighttpd-send-file",
  "x-sendfile",
]);

const CDN_CACHE_HEADERS = new Set([
  "cdn-cache-control",
  "cloudflare-cdn-cache-control",
  "netlify-cdn-cache-control",
  "surrogate-control",
]);

const BASIC_AUTH_CREDENTIAL_PATTERN = /^[^:\s]+:[^\s]+$/;
const CARRIAGE_RETURN_PATTERN = /\r$/;
const HEADER_NAME_PATTERN = /^[A-Za-z0-9-]+$/;
const LINE_SPLIT_PATTERN = /\r?\n/;

function canonicalHeaderName(input: string) {
  return input
    .trim()
    .toLowerCase()
    .split("-")
    .map((part) => (part ? `${part[0]?.toUpperCase() ?? ""}${part.slice(1)}` : part))
    .join("-");
}

function compileHeaderMatcher(
  rawPath: string,
  line: number,
  sourceLine: string,
  diagnostics: RoutingDiagnostic[],
) {
  const absolute = parseAbsoluteUrl(rawPath);
  if (absolute) {
    if (absolute.port) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_port_unsupported",
        "Header URL matchers cannot include ports.",
        sourceLine,
      );
      return null;
    }

    const host = normalizeHostname(absolute.host);
    const hostCompiled = compilePattern(host, ".", {
      allowWildcardPrefixInSegment: true,
    });
    const pathCompiled = compilePattern(absolute.path, "/", {
      allowWildcardPrefixInSegment: true,
    });
    if (hostCompiled.error || pathCompiled.error) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_pattern_invalid",
        hostCompiled.error ?? pathCompiled.error ?? "This header matcher is not valid.",
        sourceLine,
      );
      return null;
    }

    return {
      path: absolute.path,
      host,
      regex: pathCompiled.regex ?? `^${escapeRegex(absolute.path)}$`,
      hostRegex: hostCompiled.regex ?? `^${escapeRegex(host)}$`,
    };
  }

  if (!rawPath.startsWith("/")) {
    diagnostic(
      diagnostics,
      "_headers",
      line,
      "error",
      "header_path_invalid",
      "The header matcher must start with / or use an absolute URL.",
      sourceLine,
    );
    return null;
  }

  const compiled = compilePattern(rawPath, "/", {
    allowWildcardPrefixInSegment: false,
  });
  if (compiled.error) {
    diagnostic(
      diagnostics,
      "_headers",
      line,
      "error",
      "header_pattern_invalid",
      compiled.error,
      sourceLine,
    );
    return null;
  }

  return {
    path: rawPath,
    host: null,
    regex: compiled.regex ?? `^${escapeRegex(rawPath)}$`,
    hostRegex: null,
  };
}

function normalizeHeaderOperation(
  operation: CompiledHeaderOperation,
  line: number,
  sourceLine: string,
  diagnostics: RoutingDiagnostic[],
) {
  const name = operation.name.trim();
  const lowerName = name.toLowerCase();

  if (lowerName === "basic-auth") {
    if (operation.kind !== "set" || !operation.value?.trim()) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_basic_auth_invalid",
        "Basic-Auth needs one or more username:password credentials.",
        sourceLine,
      );
      return null;
    }

    const credentials = operation.value.trim().split(WHITESPACE_PATTERN);
    if (credentials.some((credential) => !BASIC_AUTH_CREDENTIAL_PATTERN.test(credential))) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_basic_auth_invalid",
        "Basic-Auth credentials must use username:password.",
        sourceLine,
      );
      return null;
    }

    diagnostic(
      diagnostics,
      "_headers",
      line,
      "warning",
      "header_basic_auth_never_emitted",
      "Basic-Auth protects matching paths and is not sent as a response header.",
      sourceLine,
    );
    return { kind: "basicAuth" as const, name: "Basic-Auth", value: operation.value.trim() };
  }

  if (!HEADER_NAME_PATTERN.test(name)) {
    diagnostic(
      diagnostics,
      "_headers",
      line,
      "error",
      "header_name_invalid",
      "This header name is not valid.",
      sourceLine,
    );
    return null;
  }

  if (UNSUPPORTED_RESPONSE_HEADERS.has(lowerName)) {
    diagnostic(
      diagnostics,
      "_headers",
      line,
      "error",
      "header_name_unsupported",
      `The "${canonicalHeaderName(name)}" header is managed by the platform and cannot be changed.`,
      sourceLine,
    );
    return null;
  }

  if (CDN_CACHE_HEADERS.has(lowerName)) {
    diagnostic(
      diagnostics,
      "_headers",
      line,
      "error",
      "header_cdn_cache_unsupported",
      `The "${canonicalHeaderName(name)}" header does not have Spacefast semantics.`,
      sourceLine,
    );
    return null;
  }

  if (lowerName === "cache-control") {
    diagnostic(
      diagnostics,
      "_headers",
      line,
      "warning",
      "header_cache_control_platform_managed",
      "Cache-Control applies to browser responses; platform edge caching is managed by Spacefast.",
      sourceLine,
    );
  }

  return { ...operation, name: canonicalHeaderName(name) };
}

export function compileHeaders(content: string, diagnostics: RoutingDiagnostic[]) {
  const rules: CompiledHeaderRule[] = [];
  let currentPath: { path: string; line: number; source: string } | null = null;
  let operations: Array<CompiledHeaderOperation & { line: number; source: string }> = [];

  const flush = () => {
    if (!currentPath) {
      operations = [];
      return;
    }

    const matcher = compileHeaderMatcher(
      currentPath.path,
      currentPath.line,
      currentPath.source,
      diagnostics,
    );
    const normalizedOperations = operations.flatMap((operation) => {
      const normalized = normalizeHeaderOperation(
        operation,
        operation.line,
        operation.source,
        diagnostics,
      );
      return normalized ? [normalized] : [];
    });

    if (matcher && normalizedOperations.length > 0) {
      const headers: Record<string, string> = {};
      for (const operation of normalizedOperations) {
        if (operation.kind === "set" && operation.value !== null) {
          headers[operation.name] = operation.value;
        }
      }
      rules.push({ ...matcher, operations: normalizedOperations, headers });
    }

    currentPath = null;
    operations = [];
  };

  for (const [index, rawLine] of content.split(LINE_SPLIT_PATTERN).entries()) {
    const line = index + 1;
    const sourceLine = rawLine.replace(CARRIAGE_RETURN_PATTERN, "");
    const trimmed = sourceLine.trim();

    if (sourceLine.length > HEADER_LINE_LIMIT) {
      if (!sourceLine.startsWith(" ") && !sourceLine.startsWith("\t")) {
        flush();
      }
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_line_too_long",
        "This header line is too long.",
        trimmed,
      );
      continue;
    }

    if (!trimmed || trimmed.startsWith("#")) {
      continue;
    }

    if (!sourceLine.startsWith(" ") && !sourceLine.startsWith("\t")) {
      flush();
      currentPath = { path: trimmed, line, source: trimmed };
      continue;
    }

    if (!currentPath) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_operation_without_matcher",
        "Add a path or URL before this header.",
        trimmed,
      );
      continue;
    }

    if (trimmed.startsWith("!")) {
      const name = trimmed.slice(1).trim();
      if (!name) {
        diagnostic(
          diagnostics,
          "_headers",
          line,
          "error",
          "header_remove_invalid",
          "Choose a header to remove.",
          trimmed,
        );
        continue;
      }
      operations.push({ kind: "remove", name, value: null, line, source: trimmed });
      continue;
    }

    const separatorIndex = trimmed.indexOf(":");
    if (separatorIndex === -1) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_operation_invalid",
        "Header operations must use Name: value.",
        trimmed,
      );
      continue;
    }

    const name = trimmed.slice(0, separatorIndex).trim();
    const value = trimmed.slice(separatorIndex + 1).trim();
    if (!name) {
      diagnostic(
        diagnostics,
        "_headers",
        line,
        "error",
        "header_name_missing",
        "Add a header name.",
        trimmed,
      );
      continue;
    }
    operations.push({ kind: "set", name, value, line, source: trimmed });
  }

  flush();
  return rules;
}

File: crates/stattic-runtime-core/src/routing/headers.rs
//! `_headers` block grammar: matcher lines, set/remove operations, CSP and
//! platform-managed header policy (first-wins semantics are enforced by the
//! shared `crate::policy::platform_managed_response_header` list).

use crate::policy::{never_removable_response_header, platform_managed_response_header};

use super::{
    canonical_header_name, cdn_headers, compile_pattern, declaration_length, diagnostic,
    escape_regex_literal, header_name_regex, normalize_hostname, normalize_path,
    parse_absolute_url, routing_trim, HeaderOperation, HeaderRule, RoutingDiagnostic, RuleField,
    RuleIssue, HEADER_LINE_LIMIT,
};

#[derive(Debug)]
struct HeaderSourceOperation {
    operation: HeaderOperation,
    line: usize,
    source: String,
}

pub(super) type HeaderMatcher = (String, Option<String>, Option<String>, Option<String>);

/// What the shared header-name policy decided. Callers attach the location: a
/// `_headers` line, or a `sf.jsonc` key path.
pub(super) enum HeaderNameVerdict {
    Accepted {
        canonical: String,
        advisory: Option<HeaderNameNote>,
    },
    Rejected(HeaderNameNote),
}

pub(super) struct HeaderNameNote {
    pub code: &'static str,
    pub message: String,
}

pub(super) fn compile_headers(
    content: &str,
    diagnostics: &mut Vec<RoutingDiagnostic>,
) -> Vec<HeaderRule> {
    let mut rules = Vec::new();
    let mut current: Option<(String, usize, String)> = None;
    let mut operations: Vec<HeaderSourceOperation> = Vec::new();
    let flush = |current: &mut Option<(String, usize, String)>,
                 operations: &mut Vec<HeaderSourceOperation>,
                 rules: &mut Vec<HeaderRule>,
                 diagnostics: &mut Vec<RoutingDiagnostic>| {
        let Some((path, line, source)) = current.take() else {
            operations.clear();
            return;
        };
        let matcher = match compile_header_matcher(&path) {
            Ok(matcher) => Some(matcher),
            Err(issue) => {
                diagnostic(
                    diagnostics,
                    "_headers",
                    line,
                    issue.severity,
                    issue.code,
                    &issue.message,
                    &source,
                );
                None
            }
        };
        let normalized: Vec<HeaderOperation> = operations
            .drain(..)
            .filter_map(|item| normalize_header_operation(item, diagnostics))
            .collect();
        if let Some((path, host, regex, host_regex)) = matcher.filter(|_| !normalized.is_empty()) {
            let headers = normalized
                .iter()
                .filter(|op| op.kind == "set")
                .map(|op| (op.name.clone(), op.value.clone().unwrap_or_default()))
                .collect();
            rules.push(HeaderRule {
                path,
                host,
                regex,
                host_regex,
                operations: normalized,
                headers,
                origin: "file",
            });
        }
    };
    for (index, raw_line) in content.lines().enumerate() {
        let line = index + 1;
        // One trailing CR, the way `/\r$/` strips it — not every trailing CR.
        let source_line = raw_line.strip_suffix('\r').unwrap_or(raw_line);
        let trimmed = routing_trim(source_line);
        if declaration_length(source_line) > HEADER_LINE_LIMIT {
            if !source_line.starts_with([' ', '\t']) {
                flush(&mut current, &mut operations, &mut rules, diagnostics);
            }
            diagnostic(
                diagnostics,
                "_headers",
                line,
                "error",
                "header_line_too_long",
                "This header line is too long.",
                trimmed,
            );
            continue;
        }
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        if !source_line.starts_with([' ', '\t']) {
            flush(&mut current, &mut operations, &mut rules, diagnostics);
            current = Some((trimmed.to_string(), line, trimmed.to_string()));
            continue;
        }
        if current.is_none() {
            diagnostic(
                diagnostics,
                "_headers",
                line,
                "error",
                "header_operation_without_matcher",
                "Add a path or URL before this header.",
                trimmed,
            );
            continue;
        }
        if let Some(name) = trimmed.strip_prefix('!') {
            if routing_trim(name).is_empty() {
                diagnostic(
                    diagnostics,
                    "_headers",
                    line,
                    "error",
                    "header_remove_invalid",
                    "Choose a header to remove.",
                    trimmed,
                );
            } else {
                operations.push(HeaderSourceOperation {
                    operation: HeaderOperation {
                        kind: "remove",
                        name: routing_trim(name).to_string(),
                        value: None,
                        line: None,
                        source: None,
                    },
                    line,
                    source: trimmed.to_string(),
                });
            }
            continue;
        }
        let Some((name, value)) = trimmed.split_once(':') else {
            diagnostic(
                diagnostics,
                "_headers",
                line,
                "error",
                "header_operation_invalid",
                "Header operations must use Name: value.",
                trimmed,
            );
            continue;
        };
        if routing_trim(name).is_empty() {
            diagnostic(
                diagnostics,
                "_headers",
                line,
                "error",
                "header_name_missing",
                "Add a header name.",
                trimmed,
            );
            continue;
        }
        operations.push(HeaderSourceOperation {
            operation: HeaderOperation {
                kind: "set",
                name: routing_trim(name).to_string(),
                value: Some(routing_trim(value).to_string()),
                line: None,
                source: None,
            },
            line,
            source: trimmed.to_string(),
        });
    }
    flush(&mut current, &mut operations, &mut rules, diagnostics);
    rules
}

/// The header matcher grammar shared by `_headers` blocks and `sf.jsonc`
/// header rules: one matcher syntax, one set of diagnostic codes.
pub(super) fn compile_header_matcher(path: &str) -> Result<HeaderMatcher, RuleIssue> {
    if let Some(url) = parse_absolute_url(path) {
        if !url.port.is_empty() {
            return Err(RuleIssue::error(
                RuleField::Source,
                "header_port_unsupported",
                "Header URL matchers cannot include ports.",
            ));
        }
        let host = normalize_hostname(&url.host);
        let normalized_path = normalize_path(&url.path);
        let host_compiled = compile_pattern(&host, '.', true, false);
        let path_compiled = compile_pattern(&normalized_path, '/', true, false);
        if host_compiled.error.is_some() || path_compiled.error.is_some() {
            return Err(RuleIssue::error(
                RuleField::Source,
                "header_pattern_invalid",
                host_compiled
                    .error
                    .as_deref()
                    .or(path_compiled.error.as_deref())
                    .unwrap_or("This header matcher is not valid."),
            ));
        }
        return Ok((
            normalized_path.clone(),
            Some(host.clone()),
            Some(
                path_compiled
                    .regex
                    .unwrap_or_else(|| format!("^{}$", escape_regex_literal(&normalized_path))),
            ),
            Some(
                host_compiled
                    .regex
                    .unwrap_or_else(|| format!("^{}$", escape_regex_literal(&host))),
            ),
        ));
    }
    if !path.starts_with('/') {
        return Err(RuleIssue::error(
            RuleField::Source,
            "header_path_invalid",
            "The header matcher must start with / or use an absolute URL.",
        ));
    }
    let normalized_path = normalize_path(path);
    let compiled = compile_pattern(&normalized_path, '/', false, false);
    if let Some(error) = compiled.error {
        return Err(RuleIssue::error(
            RuleField::Source,
            "header_pattern_invalid",
            error,
        ));
    }
    Ok((
        normalized_path.clone(),
        None,
        Some(
            compiled
                .regex
                .unwrap_or_else(|| format!("^{}$", escape_regex_literal(&normalized_path))),
        ),
        None,
    ))
}

/// Which header names a rule may set, shared by both grammars: Basic-Auth is
/// never a response header, names are syntax-checked, platform-managed headers
/// stay ours, and Cache-Control only advises.
pub(super) fn check_header_name(name: &str) -> HeaderNameVerdict {
    let trimmed = routing_trim(name);
    let lower = trimmed.to_lowercase();
    if lower == "basic-auth" {
        return HeaderNameVerdict::Rejected(HeaderNameNote {
            code: "header_basic_auth_unsupported",
            message:
                "Basic-Auth is not a response header. Use Spacefast sharing to control access."
                    .into(),
        });
    }
    if !header_name_regex().is_match(trimmed) {
        return HeaderNameVerdict::Rejected(HeaderNameNote {
            code: "header_name_invalid",
            message: "This header name is not valid.".into(),
        });
    }
    let canonical = canonical_header_name(trimmed);
    if platform_managed_response_header(&lower) {
        return HeaderNameVerdict::Rejected(if cdn_headers().contains(lower.as_str()) {
            HeaderNameNote {
                code: "header_cdn_cache_unsupported",
                message: format!("The \"{canonical}\" header does not have Spacefast semantics."),
            }
        } else {
            HeaderNameNote {
                code: "header_name_unsupported",
                message: format!(
                    "The \"{canonical}\" header is managed by the platform and cannot be changed."
                ),
            }
        });
    }
    HeaderNameVerdict::Accepted {
        canonical,
        advisory: (lower == "cache-control").then(|| HeaderNameNote {
            code: "header_cache_control_platform_managed",
            message: "Cache-Control applies to browser responses; platform edge caching is managed by Spacefast.".into(),
        }),
    }
}

fn normalize_header_operation(
    item: HeaderSourceOperation,
    diagnostics: &mut Vec<RoutingDiagnostic>,
) -> Option<HeaderOperation> {
    let mut operation = item.operation;
    match check_header_name(&operation.name) {
        HeaderNameVerdict::Rejected(note) => {
            diagnostic(
                diagnostics,
                "_headers",
                item.line,
                "error",
                note.code,
                &note.message,
                // A rejected Basic-Auth line carries credentials; echo the
                // line back for every other name, never for that one.
                if note.code == "header_basic_auth_unsupported" {
                    "[redacted]"
                } else {
                    &item.source
                },
            );
            None
        }
        HeaderNameVerdict::Accepted {
            canonical,
            advisory,
        } => {
            // Setting one is publisher metadata; deleting the platform's own is
            // not a rule we can honor (see `never_removable_response_header`).
            if operation.kind == "remove" && never_removable_response_header(&canonical) {
                diagnostic(
                    diagnostics,
                    "_headers",
                    item.line,
                    "error",
                    "header_name_unsupported",
                    &format!(
                        "The \"{canonical}\" header is set by Spacefast for every response and cannot be removed."
                    ),
                    &item.source,
                );
                return None;
            }
            if let Some(note) = advisory {
                diagnostic(
                    diagnostics,
                    "_headers",
                    item.line,
                    "warning",
                    note.code,
                    &note.message,
                    &item.source,
                );
            }
            operation.name = canonical;
            operation.line = Some(item.line);
            operation.source = Some(item.source);
            Some(operation)
        }
    }
}
